File: src/stores/folders.ts
// stores/folders.ts
import { defineStore } from "pinia";
import type { Folder } from "../types/folder";
import {
  foldersList,
  foldersCreate,
  foldersRename,
  foldersDelete,
  foldersReorder,
} from "../api/folders";

interface State {
  list: Folder[];
  loaded: boolean;
}

export const useFoldersStore = defineStore("folders", {
  state: (): State => ({ list: [], loaded: false }),
  actions: {
    async loadAll(): Promise<void> {
      this.list = await foldersList();
      this.loaded = true;
    },
    async create(name: string): Promise<Folder> {
      const f = await foldersCreate(name);
      await this.loadAll();
      return f;
    },
    async rename(id: number, name: string): Promise<void> {
      await foldersRename(id, name);
      await this.loadAll();
    },
    async remove(id: number): Promise<void> {
      await foldersDelete(id);
      await this.loadAll();
    },
    async reorder(orderedIds: number[]): Promise<void> {
      await foldersReorder(orderedIds);
      await this.loadAll();
    },
  },
});

File: src-tauri/src/commands/folders.rs
// commands/folders.rs
use tauri::{AppHandle, State};

use crate::db::{self, DbState};
use crate::error::AppResult;
use crate::events;
use crate::models::folder::Folder;

#[tauri::command]
pub fn folders_list(db: State<'_, DbState>) -> AppResult<Vec<Folder>> {
    let conn = db.0.lock();
    db::folders::list(&conn)
}

#[tauri::command]
pub fn folders_create(app: AppHandle, db: State<'_, DbState>, name: String) -> AppResult<Folder> {
    let f = {
        let conn = db.0.lock();
        db::folders::create(&conn, &name)?
    };
    events::emit_folders_changed(&app);
    Ok(f)
}

#[tauri::command]
pub fn folders_rename(
    app: AppHandle,
    db: State<'_, DbState>,
    id: i64,
    name: String,
) -> AppResult<()> {
    {
        let conn = db.0.lock();
        db::folders::rename(&conn, id, &name)?;
    }
    events::emit_folders_changed(&app);
    Ok(())
}

#[tauri::command]
pub fn folders_delete(app: AppHandle, db: State<'_, DbState>, id: i64) -> AppResult<()> {
    {
        let conn = db.0.lock();
        db::folders::delete(&conn, id)?;
    }
    events::emit_folders_changed(&app);
    events::emit_prompts_changed(&app); // 关联 prompt 的 folder_id 被置 NULL
    Ok(())
}

#[tauri::command]
pub fn folders_reorder(
    app: AppHandle,
    db: State<'_, DbState>,
    ordered_ids: Vec<i64>,
) -> AppResult<()> {
    {
        let mut conn = db.0.lock();
        db::folders::reorder(&mut conn, &ordered_ids)?;
    }
    events::emit_folders_changed(&app);
    Ok(())
}
